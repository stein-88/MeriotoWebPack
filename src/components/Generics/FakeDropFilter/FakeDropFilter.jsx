import React, { useState, useEffect, useRef } from 'react'
import GenInput from '../GenInput'
import s from './FakeDropFilter.scss'

const FakeDropFilter = ({
	classItems,
	classBox,
	trocaTexto,
	listaSchema,
	value,
	customClass,
	id,
	placeholder,
	onChange,
}) => {
	const [mostrardrop, setMostrarDrop] = useState(false)
	const didMount = useRef(true)
	const refElement = useRef()
	const dropRef = useRef(mostrardrop)

	const changeDropRef = (param) => {
		dropRef.current = param
		return setMostrarDrop(param)
	}

	const outClick = (e) => {
		if (!e || !e.target) return null
		if (dropRef.current && refElement.current && !refElement.current.contains(e.target)) {
			return changeDropRef(false)
		}
		return null
	}

	useEffect(() => {
		document.body.addEventListener('mousedown', outClick)
		return () => {
			document.body.removeEventListener('mousedown', outClick)
			didMount.current = false
		}
	}, [])

	if (!listaSchema || !listaSchema.length || !id || !trocaTexto || !onChange) return null
	const resultado = listaSchema.filter((ele) => ele.toLowerCase().includes(value.toLowerCase()))
	return (
		<section ref={refElement} className="w-100">
			<GenInput
				type="text"
				placeholder={placeholder}
				value={value || ''}
				id={id}
				onChange={onChange}
				className={`w-100 ${s.classedrop || customClass}`}
				onFocus={() => {
					if (mostrardrop) return null
					return changeDropRef(true)
				}}
			/* onBlur={() => setTimeout(() => {
				if (didMount.current) return setMostrarDrop(false)
				return null
			}, 1000)} */
			/>
			{mostrardrop
				&& (
					<section className={s.caixinha || classBox}>
						{resultado && resultado.length > 0
							? resultado.map((ele) => (
								<p
									key={ele + 1}
									onClick={() => {
										trocaTexto(ele)
										return changeDropRef(false)
									}}
									aria-hidden="true"
									className={s.item || classItems}
								>
									{ele}
								</p>
							))
							: <p>Nenhum resultado encontrado</p>}
					</section>
				)}
		</section>
	)
}

export default FakeDropFilter
