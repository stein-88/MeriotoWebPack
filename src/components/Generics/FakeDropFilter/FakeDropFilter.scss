.classedrop {
  border: 1px solid black;
  border-radius: 22px;
  padding: 15px;
  cursor: pointer;
}

.caixinha {
  border: 1px solid black;
  border-top: none;
  border-radius: 0 0 22px 22px;
  overflow: scroll;
  max-height: 200px;

  .item {
    padding: 15px;
    margin-bottom: 0;
    cursor: pointer;

    &:hover {
      background-color: aquamarine;
    }
  }
}
