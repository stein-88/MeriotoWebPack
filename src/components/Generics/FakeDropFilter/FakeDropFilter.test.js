import React from 'react'
import renderer from 'react-test-renderer'
import FakeDropFilter from './FakeDropFilter'

describe('Test on FakeDropFilter Component', () => {
    const compFakeDropFilter = renderer.create(<FakeDropFilter />)
    test('Should not render without props', () => {
        const jsonFakeDropFilter = compFakeDropFilter.toJSON()
        expect(jsonFakeDropFilter).toBeNull()
    })
    test('Should not render with 1 props only', () => {
        const compListaSchema = renderer.create(<FakeDropFilter listaSchema={
            [
                'legume',
                'verdura',
                'fruta'
            ]
        } />)
        const compTroca = renderer.create(<FakeDropFilter trocaTexto={() => null} />)
        const componChange = renderer.create(<FakeDropFilter onChange={() => null} />)
        const compID = renderer.create(<FakeDropFilter ID={'banana'} />)
        const jsonListaSchema = compListaSchema.toJSON()
        const jsonTroca = compTroca.toJSON()
        const jsononChange = componChange.toJSON()
        const jsonID = compID.toJSON()
        expect(jsonListaSchema).toBeNull()
        expect(jsonTroca).toBeNull()
        expect(jsononChange).toBeNull()
        expect(jsonID).toBeNull()
    })
    test('Should not render with 2 props', () => {
        const compFakeDropFilter = renderer.create(<FakeDropFilter ListaSchema={
            [
                'legume',
                'verdura',
                'fruta'
            ]
        } trocaTexto={() => null} />)
        const jsonFakeDropFilter = compFakeDropFilter.toJSON()
        expect(jsonFakeDropFilter).toBeNull()
    })
    test('Should not render with 3 props', () => {
        const compFakeDropFilter = renderer.create(<FakeDropFilter
            ListaSchema={
                [
                    'legume',
                    'verdura',
                    'fruta'
                ]
            }
            trocaTexto={() => null}
            ID={'banana'}
        />)
        const jsonFakeDropFilter = compFakeDropFilter.toJSON()
        expect(jsonFakeDropFilter).toBeNull()
    })
    test('Should render with full props', () => {
        const compFakeDropFilter = renderer.create(<FakeDropFilter
            ListaSchema={
                [
                    'legume',
                    'verdura',
                    'fruta'
                ]
            }
            trocaTexto={() => null}
            ID={'banana'}
            onChange={() => null}
        />)
        const jsonFakeDropFilter = compFakeDropFilter.toJSON()
        expect(jsonFakeDropFilter).toMatchSnapshot()
    })
})