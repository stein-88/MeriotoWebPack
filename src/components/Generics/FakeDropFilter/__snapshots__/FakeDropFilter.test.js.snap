// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on FakeDropFilter Component Should render with full props 1`] = `null`;
